src/app/cli/src/services/odf_server/login_service.rs

```diff
@@ -263,9 +241,103 @@ impl LoginService {
                     odf_server_backend_url: odf_server_backend_url.clone(),
                 }))
             }
-            _ => panic!(
+            unexpected_status => panic!(
                 "Unexpected validation status code: {}, details: {}",
-                response.status().as_str(),
+                unexpected_status.as_u16(),
+                response.text().await.unwrap()
+            ),
+        }
+    }
+
+    async fn odf_server_backend_url(
+        &self,
+        client: &reqwest::Client,
+        odf_server_frontend_url: &Url,
+    ) -> Result<Url, InternalError> {
+        let response = client
+            .get(
+                odf_server_frontend_url
+                    .join("assets/runtime-config.json")
+                    .unwrap(),
+            )
+            .send()
+            .await
+            .int_err()?;
+
+        #[derive(Debug, Deserialize)]
+        #[serde(rename_all = "camelCase")]
+        struct Response {
+            api_server_http_url: Url,
+        }
+
+        match response.status() {
+            http::StatusCode::OK => Ok(response
+                .json::<Response>()
+                .await
+                .int_err()?
+                .api_server_http_url),
+            unexpected_status => panic!(
+                "Unexpected runtime-config status code: {}, details: {}",
+                unexpected_status.as_u16(),
+                response.text().await.unwrap()
+            ),
+        }
+    }
+
+    async fn device_authorization(
+        &self,
+        client: &reqwest::Client,
+        odf_server_backend_url: &Url,
+    ) -> Result<DeviceAuthorizationResponse, InternalError> {
+        let response = client
+            .post(
+                odf_server_backend_url
+                    .join("platform/token/device/authorization")
+                    .unwrap(),
+            )
+            // We do not use this value, but we are required to pass it.
+            .form(&[("client_id", "kamu")])
+            .send()
+            .await
+            .int_err()?;
+
+        match response.status() {
+            http::StatusCode::OK => Ok(response.json().await.int_err()?),
+            unexpected_status => panic!(
+                "Unexpected device authorization status code: {}, details: {}",
+                unexpected_status.as_u16(),
+                response.text().await.unwrap()
+            ),
+        }
+    }
+    async fn device_access_token(
+        &self,
+        client: &reqwest::Client,
+        login_polling_url: &Url,
+        device_code: &str,
+    ) -> Result<DeviceAccessTokenResult, InternalError> {
+        let response = client
+            .post(login_polling_url.clone())
+            .form(&[
+                // We do not use this value, but we are required to pass it.
+                ("client_id", "kamu"),
```

Comment:
It's used twice at least, so make a constant

Reply:
Done